homie-controller/src/lib.rs

```diff
@@ -248,6 +252,82 @@ impl HomieController {
                 device.implementation = Some(payload.to_owned());
                 Some(Event::device_updated(device))
             }
+            [device_id, "$extensions"] => {
+                let device = get_mut_device_for(devices, "Got extensions for", device_id)?;
+                device.extensions = payload
+                    .split(",")
+                    .map(|part| part.parse())
+                    .collect::<Result<Vec<_>, _>>()?;
+                Some(Event::device_updated(device))
+            }
+            [device_id, "$localip"] => {
+                let device = get_mut_device_for(devices, "Got localip for", device_id)?;
+                device.local_ip = Some(payload.to_owned());
+                Some(Event::device_updated(device))
+            }
+            [device_id, "$mac"] => {
+                let device = get_mut_device_for(devices, "Got mac for", device_id)?;
+                device.mac = Some(payload.to_owned());
+                Some(Event::device_updated(device))
+            }
+            [device_id, "$fw", "name"] => {
```

Comment:
Might be clearer to use `let parts = subtopic.splitn("/", 2).collect::<Vec<&str>>();` above, and `[device_id, "$fw/name"]` here, now that this is the majority case.

I would also start thinking about using a macro at this point. Would it be possible to write this as:

```
event_case!("$fw/name", firmware_name, std::borrow::ToOwned::to_owned)
```
 or something? (it might not be possible. I have no idea).

Reply:
How would that work for the node and property attributes below?